src/kudu/consensus/leader_election.cc

```diff
@@ -1097,9 +1124,37 @@ std::pair<bool, bool> FlexibleVoteCounter::IsDynamicQuorumSatisfied() const {
             .count();
     if (pessimistic_result.second &&
         time_elapsed_secs < FLAGS_wait_for_pessimistic_quorum_secs) {
+      LOG_WITH_PREFIX(INFO) << "Pausing for Pessimistic quorum to help decide election";
+      return pessimistic_result;
+    }
+
+    // Set this to a very high value if we don't want to fallback to
+    // Voting history, which has some known gaps in its logic.
+    // Risk with that is when there is a single region failure, pessimistic
+    // quorum will return failure but the election can still be won by finding
+    // a smaller set of previous leader regions to intersect with.
+    //
+    // This code path is only taken when there is a discontinuity
+    // in election terms, it is rare (since we run pre-elections
+    // for Dead Primary Promotions before term bump). Today Live Primary
+    // Promotions, Race of 2 CANDIDATEs or Operator introduced Promotions
+    // are the only cases where a term discontinuity can happen.
+    if (FLAGS_wait_before_using_voting_history &&
+        time_elapsed_secs < FLAGS_wait_before_using_voting_history) {
+      LOG_WITH_PREFIX(INFO) << "Pessimistic quorum did not help decide election but pausing for: "
+          << (FLAGS_wait_before_using_voting_history - time_elapsed_secs)
+          << " seconds before trying voting history heuristic";
       return pessimistic_result;
     }
 
+    LOG_WITH_PREFIX(INFO) << "Using Voting History Fallback due to term discontinuity"
+        << " Election term: " << election_term_
+        << " lkl term: " << last_known_leader.election_term()
+        << " lkl uuid: " << last_known_leader.uuid()
+        << " lkl region: " << last_known_leader_region
+        << " continuity_not_required: " << continuity_not_required
+        << " is_continuous: " << is_continuous;
+
     // Step 4.2: We come here if either pessimistic quorum satisfaction is not
```

Comment:
nit: This comment needs an update

Also maybe we should print all the potential leaders found using voter history in ComputeElectionResultFromVotingHistory()?